src/plugins/charset_converter.c

```diff
@@ -0,0 +1,185 @@
+#include "plugins/charset_converter.h"
+#include "flutter-pi.h"
+#include "pluginregistry.h"
+#include "util/logging.h"
+#include <iconv.h>
+
+static bool convert(char *buf, char *outbuf, size_t len, const char *from, const char *to)
+{
+    iconv_t iconv_cd;
+    if ((iconv_cd = iconv_open(to, from)) == (iconv_t) -1) {
+        LOG_ERROR("Cannot open iconv from %s to %s\n", from, to);
+        return false;
+    }
+
+    char *inbuf = buf;
+    size_t inlen = len;
+    size_t outlen = len;
+    size_t res = 0;
+
+    while (inlen > 0 && outlen > 0) {
+        res = iconv(iconv_cd, &inbuf, &inlen, &outbuf, &outlen);
+        if (res == 0)
+            break;
+
+        if (res == (size_t) (-1))
+        {
+            if (errno != EILSEQ && errno != EINVAL)
+            {
+                iconv_close(iconv_cd);
+                *outbuf = '\0';
+
+                return true;
+            }
+            else if (inbuf < outbuf) {
+                iconv_close(iconv_cd);
+                *outbuf = '\0';
+
+                return true;
+            }
+        }
+
+        if (inlen > 0 && outlen > 0) {
```

Comment:
Does this case ever happen? Why are you copying over a single unconverted character here?